src/plugin/sdl/sdl.c

```diff
@@ -1019,6 +1179,182 @@ static void sdl_scrub(void)
+  SDL_RenderCopy(renderer, txt, NULL, &rect);
+  pthread_mutex_unlock(&rend_mtx);
+
+  pthread_mutex_lock(&rects_mtx);
+  sdl_rects_num++;
+  pthread_mutex_unlock(&rects_mtx);
+
+  SDL_DestroyTexture(txt);
+}
+
+/*
+ * Draw a horizontal line (for text modes)
+ * The attribute is the VGA color/mono text attribute.
+ */
+static void SDL_draw_line(void *opaque, int x, int y, int len)
+{
+  v_printf("SDL_draw_line x(%d) y(%d) len(%d)\n", x, y, len);
+
+  pthread_mutex_lock(&rend_mtx);
+  SDL_RenderDrawLine(renderer,
+      font_width * x,
+      font_height * y,
+      font_width * (x + len) - 1,
+      font_height * y
+  );
+  pthread_mutex_unlock(&rend_mtx);
+
+  pthread_mutex_lock(&rects_mtx);
+  sdl_rects_num++;
+  pthread_mutex_unlock(&rects_mtx);
+}
+
+/*
+ * Draw the cursor (nothing in graphics modes, normal if we have focus,
+ * rectangle otherwise).
+ */
+static void SDL_draw_text_cursor(void *opaque, int x, int y, Bit8u attr,
+                               int start, int end, Boolean focus)
+{
+  SDL_Rect rect;
+
+  if (vga.mode_class == GRAPH)
+    return;
+
+  if (!focus) {
+    rect.x = font_width * x;
+    rect.y = font_height * y;
+    rect.w = font_width;
+    rect.h = font_height - 1;
+
+    pthread_mutex_lock(&rend_mtx);
+    SDL_SetRenderDrawColor(renderer,
+                           text_colors[ATTR_FG(attr)].r,
+                           text_colors[ATTR_FG(attr)].g,
+                           text_colors[ATTR_FG(attr)].b,
+                           text_colors[ATTR_FG(attr)].a);
+    SDL_RenderDrawRect(renderer, &rect);
+    pthread_mutex_unlock(&rend_mtx);
+
+  } else {
+    int cstart, cend;
+
+    cstart = ((start + 1) * font_height) / vga.char_height - 1;
+    if (cstart == -1)
+      cstart = 0;
+    cend = ((end + 1) * font_height) / vga.char_height - 1;
+    if (cend == -1)
+      cend = 0;
+
+    rect.x = font_width * x;
+    rect.y = font_height * y + cstart;
+    rect.w = font_width;
+    rect.h = cend - cstart + 1;
+
+    pthread_mutex_lock(&rend_mtx);
+    SDL_SetRenderDrawColor(renderer,
+                           text_colors[ATTR_FG(attr)].r,
+                           text_colors[ATTR_FG(attr)].g,
+                           text_colors[ATTR_FG(attr)].b,
+                           text_colors[ATTR_FG(attr)].a);
+    SDL_RenderFillRect(renderer, &rect);
+    pthread_mutex_unlock(&rend_mtx);
+  }
+
+  pthread_mutex_lock(&rects_mtx);
+  sdl_rects_num++;
+  pthread_mutex_unlock(&rects_mtx);
+}
+
+/*
+ * Update the active SDL colormap for text modes DAC entry col.
+ */
+static void SDL_set_text_palette(void *opaque, DAC_entry *col, int i)
+{
+  int shift = 8 - vga.dac.bits;
+
+  v_printf("SDL_set_text_palette %d: shift %d, rgb(%x, %x, %x)\n", i, shift, col->r << shift, col->g << shift, col->b << shift);
+
+  text_colors[i].r = col->r << shift;
+  text_colors[i].g = col->g << shift;
+  text_colors[i].b = col->b << shift;
+  text_colors[i].a = 0;
+}
+
+static int SDL_text_lock(void *opaque)
+{
```

Comment:
Empty call-backs?

Reply:
Indeed, I don't know what they do or what to do about those yet.

Reply:
See above, you need to lock surface
and then use blit.